README.md
# bibparse -- read and write BibTeX files

## Description

`bibparse` reads and writes BibTeX files.

The main class, `Biblio`, is a `dict` with methods for parsing, reading, writing and searching for BibTeX data. Each entry in the `Biblio` is another kind of special dict, `BibItem`. The user usually only needs to access `Biblio`.

**NOTE** Just because I use only `{...}` myself, I’ve never thought of adding the more traditional `"..."` field marking before version 1.2.*.

## Current version

Version 1.3.0-dev.1 -- 5 July 2023.

## Copyrights

Copyright © 2019–2023 Legisign.org <mailto:software@legisign.org>

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.  If not, see <https://www.gnu.org/licenses/>.

## Example usage

    import bibparse
    db = bibparse.Biblio(filename)
    print(db)

The sample code loads, parses and pretty-prints a BibTeX file.

## Module contents

### 1. Helper functions

The values in some fields (namely author, editor, translator, publisher, address, and pages, all of which may have several names, locations, or other data in their values) of `BibItem` are stored internally as `list`s.

* `to_bibtex(key, val)` -- convert an internal Python value into a BibTeX string
* `to_python(key, val)` -- convert a BibTeX string into an internal Python value

Both functions take a BibTeX field name (key) in order to decide how to handle the value. Except for "pages" where the separator is a single dash "-", the separator is the string " and ", the leading and trailing whitespace included.

Examples:

*   `to_bibtex('pages', [100, 110])` →  '100-110'
*   `to_python('address', 'London and New York')` → ['London', 'New York']

### 2. Exceptions

* `BibError` -- the base exception
    * `DuplicateError` -- duplicate ID’s
    * `NoIDError` -- missing ID in an entry
    * `PreambleError` -- invalid preamble

Each exception has a property `lineno` in order to refer to the line that has raised the exception:

    try:
        db = bibparse.Biblio(filename)
    except bibparse.DuplicateError as exc:
        print(f'duplicate bibid on line {exc.lineno}')

### 3. BibItem class

A `dict`-derived object representing a single BibTeX entry.

#### 3.1 Methods

These are derived from `dict` but modified to ensure lower-case keys, reasonable ordering of keys in a printout, and sort ordering.

* `__lt__()`
* `__repr__()`
* `__setitem__()`
* `gets(key)` -- get key value as BibTeX-formatted string (or the empty string)
* `update()` -- update `Biblio` using another object
* static method `parse()`: parse data into a `BibItem`.

`gets()` is for convenience: it’s equivalent to `to_bibtex(BibItem.get())`.

`update()` has an additional optional `overwrite=bool` parameter. If `True` (the default), `update()` functions exactly like `dict.update()`, updating BibItem contents from data in the supplied `dict`. If `False`, only new keys in supplied data is added but existing values are not overwritten.

### 4. Biblio class

The main class. The constructor can be given an optional filename argument; the file is opened and parsed automatically.

#### 4.1  Methods

The `__repr__()` method is provided so that merely `print()`ing the `Biblio` object produces valid BibTeX output.

* `by_bibid(bibids)` -- return all entries whose bibid is in the list (or set)
* `by_regex(field, regex)` -- search in field by regex
* `by_type(bibtypes, complement=bool)` -- search by BibTeX type
* `get(key, default)` -- return a `BibItem` or the given default (by default, `None`)
* `parse(data)`  -- parse string as BibTeX data
* `read(filename)` -- read and parse file as BibTeX data

`by_bibid(bibids)` was written in order to make combining searches easier. Each `by_regex()` call returns a `Biblio` object whose keys can be obtained with `Biblio.keys()`. These keys can be used in `set` operations to provide a new list of keys that match either any (intersection) or all (union) of the keys, and `by_bibid()` can then be used to return all the entries.

`by_regex(field, regex)` searches the database by field values and returns the matches in a new `Biblio` object. E.g., `by_regex('author', '.*Smith.*')` returns all entries where the "author" field contains "Smith".

`by_types(bibtypes, complement=False)` searches the database by BibTeX types (given without the initial `"@"`) and returns the matches in a new `Biblio` object. `bibtypes` can be a string specifying a single type (e.g., `"article"`) or a list of strings specifying several types (e.g., `["article", "book"]`). If the optional `complement` parameter is set to True, the function returns the complement, i.e., all entries _not_ matching the criteria.

`parse()` can be manually called for `str` input; however, `read()` automatically calls it, as does the constructor `Biblio()` when given a filename argument.

**NOTE**: There used to be `write(filename)` method as well, and the code for it still exists but is commented out. Usually all that is needed is to `print()` the `Biblio`.


HISTORY.md
# History and change log

## Prehistory

The project has had a long and convoluted history. My log goes as far as May 2012 when I seem to have rewrote the parsing routines from scratch. Even that was repeated in November of the same year. In June 2014 I decided to use a `dict` instead of a `list` as the main object class. The parsing routines were simplified in February 2016, and in June of that year I made major changes to them to increase their robustness.

In August 2016 I switched to using PyPI-compatible version numbering scheme in order to prepare for PyPI packaging, and made minor changes to the code in order to ensure Python 2 compatibility. The latter appears to have still caused issues since in July of the next year (2017) I had to do more changes towards it.

1 July 2019 was declared as the first “public” version (“1.0.0”), yet I continued to make bug fixes. During the next two weeks I made several semi-major changes to the code so the “1.0.0” moniker was clearly premature.

## 1.3.0 (2023-07-05)

Trying to simplify code.

## 1.2.0 (2023-04-24)

Completely rewritten parsers and lots of other internal changes, **including** the possibility to use "..." for the BibTeX fields. Python 2 compatibility hacks discarded. f-strings utilized.

## 1.1.0 (2020-01-26)

`bibkeys` updated.

## 1.0.0 (2019-07-18)

Finally there.


bibparse.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
'''bibparse -- a BibTeX parser.

  bibparse reads and writes BibTeX files.

  BibParser() is a special dict with added methods for parsing, reading,
  writing and searching for BibTeX data. Each entry in the Bibliography is
  another kind of special dict, BibEntry. Both define their own __repr__()
  methods so they can be directly printed in BibTeX format.

  Author:       Tommi Nieminen <software@legisign.org>
  License:      GPL version 3 or later

  2018-07-05  0.9.9.beta2   One more Python2 compatibility change to super().

'''

import re

version = '0.9.9.beta2'

# Recognized BibTeX keys; these keys will appear in the order given
# when BibEntry.__repr()__ is called. Any other keys in an entry will
# appear in random order tailing these ones.
bibkeys = ('key',
           'author',
           'title',
           'subtitle',
           'origtitle',
           'translator',
           'crossref',
           'editor',
           'booktitle',
           'booksubtitle',
           'origbooktitle',
           'chapter',
           'journal',
           'series',
           'volume',
           'number',
           'edition',
           'organization',
           'institution',
           'school',
           'publisher',
           'address',
           'pubstate',
           'howpublished',
           'url',
           'urldate',
           'doi',
           'month',
           'year',
           'pubyear',
           'pages',
           'isbn',
           'issn',
           'annote',
           'eprint',
           'type',
           'note',
           'keywords')

### HELPER FUNCTIONS

def to_bibtex(key, value):
    '''Convert Python value to BibTeX string.'''
    if isinstance(value, list):
        if key.lower() == 'pages':
            value = '-'.join(value)
        elif key.lower() == 'keywords':
            value = ' '.join(value)
        else:
            value = ' and '.join(value)
    return value

def to_python(key, value):
    '''Convert BibTeX string to Python value.'''
    if key.lower() in ('author', 'editor', 'publisher', 'translator'):
        value = value.split(' and ')
    elif key.lower() == 'pages':
        value = value.split('-')
    elif key.lower() == 'keywords':
        value = value.split()
    return value

### CLASSES

# Exceptions

class BibError(Exception):
    '''Base class of bibparse errors'''
    pass

class DuplicateError(BibError):
    '''Duplicate ID or preamble'''
    pass

class NoIDError(BibError):
    '''No ID in an entry'''
    pass

class PreambleError(BibError):
    '''Invalid preamble'''
    pass

# BibTeX classes

class BibEntry(dict):
    '''BibEntry is a dict containing one BibTeX entry.'''

    def __init__(self, bibid=None, bibtype=None, data=None):
        assert isinstance(data, (str, dict, type(None)))
        self.bibid = bibid
        self.bibtype = bibtype
        self.preamble = None
        if isinstance(data, dict):
            if bibtype == 'preamble':
                raise PreambleError(data)
            self.update(data)
        elif isinstance(data, str):
            if bibtype == 'preamble':
                self.preamble = data
            else:
                self.bibid, entry = BibEntry.parse(data)
                self.update(entry)

    def __lt__(self, entry):
        x = self.bibid if isinstance(self.bibid, str) else ''
        y = entry.bibid if isinstance(entry.bibid, str) else ''
        return x < y

    def __repr__(self):
        global bibkeys
        if self.bibtype == 'preamble':
            ret = '@preamble{{{}}}'.format(self.preamble)
        else:
            buff = ['@{}{{{}'.format(self.bibtype, self.bibid)] + \
                   ['    {} = {{{}}}'.format(key, to_bibtex(key, self[key])) \
                       for key in bibkeys if key in self] + \
                   ['    {} = {{{}}}'.format(key, val) \
                       for key, val in self.items() if key not in bibkeys]
            ret = ',\n'.join(buff) + '\n}'
        return ret

    def __setitem__(self, key, val):
        '''Overloaded __setitem__() to ensure lowercase keys.'''
        # super() needs arguments here for Python 2 compatibility
        super(BibEntry, self).__setitem__(key.lower(), val)

    def update(self, entries):
        '''Overloaded update() to ensure lowercase keys.'''
        # super() needs arguments here for Python 2 compatibility
        super(BibEntry, self).update({key.lower(): val \
                                     for key, val in entries.items()})

    @staticmethod
    def parse(data):
        '''Parse BibEntry data from a string.'''

        def discard_comments(s):
            '''Discard all trailing comments from data.'''
            buff = s.split('\n')
            for lineno, line in enumerate(buff):
                if '%' in line:
                    in_quotes = False
                    braces = 0
                    for col, c in enumerate(line):
                        if c == '"':
                            in_quotes = not in_quotes
                        elif c == '{':
                            braces += 1
                        elif c == '}':
                            braces -= 1
                        elif c == '%' and not in_quotes and not braces:
                            buff[lineno] = line[:col]
                            break
            return ' '.join(buff)

        data = discard_comments(data)
        bibid_pat = re.compile(r'^\s*([^,]+)\s*,\s*(.*)\s*$')
        try:
            bibid, keyvals = bibid_pat.match(data).groups()
        except AttributeError:
            raise NoIDError(data)
        bibid = bibid.strip()
        bibentry = {}
        next_is_key = True
        key = val = ''
        for c in keyvals:
            if next_is_key:
                if c.isalnum():
                    key += c
                elif c == '{':
                    next_is_key = False
                    val = c
            else:
                val += c
                if c == '}' and val.count('{') == val.count('}'):
                    bibentry[key] = to_python(key, val[1:-1])
                    next_is_key = True
                    key = val = ''
        return bibid, bibentry

class BibParser(dict):
    '''Bibliography is a dict of BibEntries.'''

    def __init__(self, filename=None, entries=None):
        self.filename = filename
        if self.filename:
            self.read(filename)
        if entries:
            if isinstance(entries, dict):
                self.update(entries)
            elif isinstance(entries, (list, set)):
                for entry in entries:
                    if not isinstance(entry, BibEntry):
                        raise ValueError(entry)
                    self[entry.bibid] = entry
            else:
                raise ValueError(entries)

    def __repr__(self):
        return '\n\n'.join([repr(entry) for entry in sorted(self.values())])

    def by_regex(self, field, pattern):
        '''Fetch all entries where field matches pattern.'''
        regex = re.compile(pattern)
        return BibParser(entries={k: v for k, v in self.items() \
                         if regex.search(str(v.get(field.lower(), '')))})

    def by_type(self, bibtypes, complement=False):
        '''Fetch all entries of given bibtype(s).'''
        if isinstance(bibtypes, str):
            bibtypes = {bibtypes}
        if not complement:
            f = lambda x, y: x in y
        else:
            f = lambda x, y: x not in y
        return BibParser(entries={k: v for k, v in self.items() \
                         if f(v.bibtype, bibtypes)})

    def parse(self, buff):
        '''Parse text buffer into a list of BibEntrys.'''
        # state can be one of the following:
        #   'wait' : wait for a bibentry (starts with “@”)
        #   'type' : read bibentry type (“book” etc.)
        #   'skip' : skip a trailing comment
        #   'data' : read bibentry data (between braces)
        state = 'wait'
        itemtype = itemdata = ''
        for c in buff:
            if state == 'skip':
                if c == '\n':
                    state = previous_state
            elif state == 'wait':
                if c == '@':
                    state = 'type'
                elif c == '{':
                    state = 'data'
                elif c == '%':
                    previous_state = state
                    state = 'skip'
            elif state == 'type':
                if c.isalpha():
                    itemtype += c.lower()
                elif c.isspace():
                    state = 'wait'
                elif c == '{':
                    state = 'data'
                elif c == '%':
                    previous_state = state
                    state = 'skip'
            elif state == 'data':
                if c == '}' and itemdata.count('{') == itemdata.count('}'):
                    if itemtype.lower() == 'preamble':
                        if '@preamble' in self:
                            raise DuplicateError
                        self['@preamble'] = BibEntry(bibtype='preamble',
                                                     data=itemdata)
                    else:
                        item = BibEntry(bibtype=itemtype, data=itemdata)
                        if item.bibid in self:
                            raise DuplicateError(item.bibid)
                        self[item.bibid] = item
                    state = 'wait'
                    itemtype = itemdata = ''
                else:
                    itemdata += c

    def read(self, filename):
        '''Read and parse a BibTeX file.'''
        self.filename = filename
        with open(filename, 'r') as f:
            buff = ' '.join([line for line in f])
        self.parse(buff)

    def write(self, filename=None, unordered=False):
        '''Write the bibliography to a BibTeX file.'''
        if not filename and not self.filename:
            raise ValueError(filename)
        if not filename:
            filename = self.filename
        with open(filename, 'w') as f:
            # Ensure decent ordering: first preamble, then not-collections,
            # then collections—because sometimes BibTeX can’t find crossrefs
            # unless they *follow* the reference
            if unordered:
                f.write(repr(self))
            else:
                f.write(repr(self.by_type('preamble')))
                f.write(repr(sorted(self.by_type(['preamble', 'collection'], \
                    complement=True))))
                f.write(repr(sorted(self.by_type('collection'))))

# Basic test if run as a script
if __name__ == '__main__':
    import sys

    def die(msg):
        # Can’t use print('…', file=…) here if called from Python 2
        sys.stderr.write(msg + '\n')
        sys.exit(1)

    for arg in sys.argv[1:]:
        try:
            db = BibParser(arg)
        except (FileNotFoundError, PermissionError, IOError):
            die('I/O-virhe: "{}"'.format(arg))
        except DuplicateError as exc:
            die('toistuva tunniste: "{}"'.format(exc.args[0]))
        except NoIDError as exc:
            die('puuttuva tunniste: "{}"'.format(exc.args[0]))
        except PreambleError:
            die('virheellinen @preamble')
        print(db)


bibparse/__init__.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-
'''bibparse -- read and write BibTeX files.

  bibparse is a BibTeX parsing module for Python.

  NOTE: Some Python 2 incompatibilities exist since 1.2.0.dev2.

  Copyright © 2019–2023 Legisign.org
  Licensed under GNU General Public License version 3 or later.

  2023-07-05  1.3.0  Trying to simplify code.

'''

import re
import enum

version = '1.3.0-dev.1'

# Recognized BibTeX keys; these keys will appear in the order given
# when BibItem.__repr()__ is called. Any other keys in an entry will
# appear in random order tailing these ones.
bibkeys = ('key',
           'author',
           'title',
           'subtitle',              # semi-standard
           'origtitle',             # non-standard
           'translator',            # semi-standard
           'crossref',
           'editor',
           'booktitle',
           'booksubtitle',          # non-standard
           'origbooktitle',         # non-standard
           'journal',
           'series',
           'volume',
           'number',
           'edition',
           'organization',
           'institution',
           'school',
           'publisher',
           'address',
           'pubstate',              # semi-standard
           'howpublished',
           'url',                   # semi-standard
           'urldate',               # semi-standard
           'doi',                   # semi-standard
           'month',
           'year',
           'pubyear',
           'chapter',
           'pages',
           'isbn',                  # semi-standard
           'issn',                  # semi-standard
           'annote',
           'type',
           'note',
           'keywords')              # semi-standard
special_types = ('comment',
                 'preamble',
                 'string')

### HELPER FUNCTIONS

def to_bibtex(key, value):
    '''Convert Python list value to BibTeX string.'''
    if isinstance(value, list):
        if key.lower() == 'pages':
            value = '-'.join(value)
        elif key.lower() == 'keywords':
            value = ' '.join(value)
        else:
            value = ' and '.join(value)
    return value

def to_python(key, value):
    '''Convert BibTeX string to a Python value.'''
    if key.lower() in ('author', 'editor', 'publisher', 'translator'):
        value = value.split(' and ')
    elif key.lower() == 'pages':
        value = value.split('-')
    elif key.lower() == 'keywords':
        value = value.split()
    return value

### CLASSES

# Exceptions

class BibError(Exception):
    '''Base class of bibparse errors'''
    def __init__(self, lineno, msg):
        self.lineno = lineno
        self.err = msg

class ParseError(BibError):
    '''Parse error.'''
    pass

class MissingCommaError(ParseError):
    '''Parser missing a comma.'''
    pass

class DuplicateError(BibError):
    '''Duplicate ID or preamble'''
    pass

class NoIDError(BibError):
    '''No ID in an entry'''
    pass

# Helper class

class ParserState(enum.Enum):
    '''File-level parser states.'''
    AT = enum.auto()        # waiting for at sign (@)
    TYPE = enum.auto()      # reading bibtype string
    SKIP = enum.auto()      # skipping characters
    DATA = enum.auto()      # waiting field (either {...} or "...")

class ItemParserState(enum.Enum):
    '''Item-level parser states.'''
    SKIP = enum.auto()      # skipping whitespace
    ID = enum.auto()        # reading bibid
    COMMA = enum.auto()     # waiting comma (or end of record)
    KEY = enum.auto()       # reading key
    EQUALS = enum.auto()    # waiting equals sign (=)
    SEPAR = enum.auto()     # waiting value separator ({ or ")
    VALUE = enum.auto()     # reading value

# BibTeX classes

class BibItem(dict):
    '''BibItem is a dict containing one BibTeX entry.'''

    def __init__(self, bibid='', bibtype='', data=None, rawdata=None):
        self.bibid = bibid
        self.bibtype = bibtype
        self.preamble = None
        self.strings = {}
        if data:
            if not isinstance(data, dict):
                raise TypeError('data should be dict')
            self.update(data)
        if rawdata:
            if not isinstance(rawdata, str):
                raise TypeError('rawdata should be str')
            if bibtype == 'preamble':
                self.preamble = rawdata
            elif bibtype == 'string':
                pass
            elif bibtype == 'comment':
                pass
            else:
                self.bibid, entry = BibItem.parse(data)
                self.update(entry)

    def __lt__(self, entry):
        x = self.bibid if isinstance(self.bibid, str) else ''
        y = entry.bibid if isinstance(entry.bibid, str) else ''
        return x < y

    def __repr__(self):
        global bibkeys
        if self.bibtype == 'preamble':
            # print(f'BibItem.__repr__(): "{self.preamble}"')
            ret = f'@preamble{{{self.preamble}}}'
        else:
            buff = [f'@{self.bibtype}{{{self.bibid}'] + \
                   [f'    {key} = {{{to_bibtex(key, self[key])}}}' \
                       for key in bibkeys if key in self] + \
                   [f'    {key} = {{{val}}}' \
                       for key, val in self.items() if key not in bibkeys]
            ret = ',\n'.join(buff) + '\n}'
        return ret

    def __setitem__(self, key, val):
        '''Overloaded __setitem__() to ensure lowercase keys.'''
        super().__setitem__(key.lower(), val)

    def gets(self, key):
        '''Get key value as a BibTeX-formatted string, or the empty string.'''
        return to_bibtex(key, self.get(key, ''))

    def update(self, fields, overwrite=True):
        '''Update item using fields, overwriting old values by default.'''
        if not isinstance(fields, dict):
            raise TypeError
        if not overwrite:
            item = {key.lower(): val for key, val in fields.items() \
                    if key.lower() not in self}
        else:
            item = fields
        super().update({key.lower(): val  for key, val in item.items()})

    @staticmethod
    def parse(data, lineno=1):
        '''Parse BibItem data from a string.'''
        state = ItemParserState.SKIP
        next_state = ItemParserState.ID
        bibid = None
        curr_key = ''
        curr_value = ''
        brackets = 0
        item = {}
        for c in data:
            # print(c, end='')
            if c == '\n':
                lineno += 1
            if state == ItemParserState.SKIP:
                if c in '{"' and next_state == ItemParserState.VALUE:
                    # print(f'SKIP -> VALUE [-> COMMA]')
                    state = ItemParserState.VALUE
                    next_state = ItemParserState.COMMA
                elif not c.isspace():
                    # print(f'*SKIP -> {next_state.name} [-> SKIP]')
                    state = next_state
                    next_state = ItemParserState.SKIP
                    curr_key = c
            elif state == ItemParserState.ID:
                if c == ',':
                    state = ItemParserState.SKIP
                    next_state = ItemParserState.KEY
                    bibid = curr_key
                    curr_key = ''
                    # print(f'ID ("{bibid}") ->> SKIP [-> KEY]')
                elif not c.isspace():
                    curr_key += c
                else:
                    state = ItemParserState.COMMA
                    next_state = ItemParserState.KEY
                    bibid = curr_key
                    curr_key = ''
                    # print(f'ID ("{bibid}") -> COMMA [-> KEY]')
            elif state == ItemParserState.COMMA:
                if c == ',':
                    # print(f'COMMA -> SKIP [-> KEY]')
                    state = ItemParserState.SKIP
                    next_state = ItemParserState.KEY
                elif not c.isspace():
                    raise MissingCommaError(lineno, c)
            elif state == ItemParserState.KEY:
                if c.isspace():
                    # print(f'KEY ("{curr_key}") -> EQUALS [-> VALUE]')
                    state = ItemParserState.EQUALS
                    next_state = ItemParserState.VALUE
                elif c == '=':
                    # print(f'KEY ("{curr_key}") ->> SKIP [-> VALUE] ')
                    state = ItemParserState.SKIP
                    next_state = ItemParserState.VALUE
                else:
                    curr_key += c
            elif state == ItemParserState.EQUALS:
                if c == '=':
                    # print(f'EQUALS -> SKIP [-> VALUE]')
                    state = ItemParserState.SKIP
                    next_state = ItemParserState.VALUE
                elif not c.isspace():
                    raise ParseError(lineno, c)
            elif state == ItemParserState.VALUE:
                if c == '{':
                    brackets += 1
                elif c == '}' and brackets > 0:
                    brackets -= 1
                elif c == '"' or (c == '}' and brackets == 0):
                    # print(f'VALUE ("{curr_value}") -> COMMA [-> KEY]')
                    state = ItemParserState.COMMA
                    next_state = ItemParserState.KEY
                    item[curr_key] = curr_value
                    curr_key = ''
                    curr_value = ''
                else:
                    curr_value += c
        return BibItem(bibid=bibid, data=item)

class Biblio(dict):
    '''Biblio is a dict of BibEntries.'''

    def __init__(self, filename=None, entries=None):
        self.filename = filename
        if self.filename:
            self.read(filename)
        if entries:
            if isinstance(entries, dict):
                self.update(entries)
            elif isinstance(entries, (list, set)):
                for entry in entries:
                    if not isinstance(entry, BibItem):
                        raise ValueError(entry)
                    self[entry.bibid] = entry
            else:
                raise ValueError(entries)

    def __repr__(self):
        # return '\n\n'.join([repr(entry) for entry in self.values()])
        return '\n\n'.join([repr(entry) for entry in sorted(self.values())])

    def by_bibid(self, bibids):
        '''Fetch all entries whose bibid is in the given list.'''
        return Biblio(entries={k: v for k, v in self.items() if k in bibids})

    def by_regex(self, field, pattern):
        '''Fetch all entries where field matches pattern (a regex).'''
        field = field.lower()
        # Search 'subtitle' field too if 'title' given as the target
        fields = [field] if field != 'title' else [field, 'subtitle']
        regex = re.compile(pattern, re.I)
        match = lambda what, where: regex.search(what.gets(where))
        results = Biblio()
        for f in fields:
            results.update({k: v for k, v in self.items() if match(v, f)})
        return results

    def by_type(self, bibtypes, complement=False):
        '''Fetch all entries of given bibtype(s).'''
        if isinstance(bibtypes, str):
            bibtypes = {bibtypes}
        if complement:
            match = lambda item: item.bibtype not in bibtypes
        else:
            match = lambda item: item.bibtype in bibtypes
        return Biblio(entries={k: v for k, v in self.items() if match(v)})

    def get(self, key, default=None):
        '''Return self[key] if key in self, else return default.

        Provides special handling for key == 'bibid'.'''
        return super().get(key, default) if key != 'bibid' else self.bibid

    def parse(self, buff):
        '''Parse text buffer into a list of BibItems.'''
        global special_types
        state = ParserState.AT
        next_state = ParserState.TYPE
        curr_type = ''
        curr_data = ''
        line = 1
        for c in buff:
            # print(c, end='')
            if c == '\n':
                line += 1
            if state == ParserState.AT and c == '@':
                # print(f'AT -> {next_state.name} [-> SKIP]')
                state = next_state
                next_state = ParserState.SKIP
            elif state == ParserState.TYPE:
                if c.isalpha():
                    curr_type += c
                elif c.isspace():
                    # print(f'TYPE ("{curr_type}") -> {next_state.name} [-> DATA]')
                    state = next_state
                    next_state = ParserState.DATA
                elif c == '{':
                    # print(f'TYPE ("{curr_type}") ->> DATA [-> AT]')
                    brackets = 1
                    state = ParserState.DATA
                    next_state = ParserState.AT
                else:
                    raise ParseError(line, c)
            elif state == ParserState.SKIP and  c == '{':
                # print(f'SKIP -> DATA [-> AT]')
                brackets = 1
                state = ParserState.DATA
                next_state = ParserState.AT
            elif state == ParserState.DATA:
                if c == '}':
                    brackets -= 1
                    if brackets == 0:
                        # print(f'DATA -> {next_state.name} [-> TYPE]')
                        # print(f'data = "{curr_data}"')
                        if curr_type not in special_types:
                            item = BibItem.parse(curr_data, line)
                            item.bibtype = curr_type
                            self[item.bibid] = item
                        elif curr_type == 'preamble':
                            # print(f'curr_data = "{curr_data}"')
                            item = BibItem(bibtype='preamble', rawdata=curr_data)
                            # print(f'item.preamble = "{item.preamble}"')
                            self[item.bibid] = item
                        state = next_state
                        next_state = ParserState.TYPE
                        curr_type = ''
                        curr_data = ''
                    else:
                        curr_data += c
                else:
                    if c == '{':
                        brackets += 1
                    curr_data += c

    def read(self, filename):
        '''Read and parse a BibTeX file.'''
        self.filename = filename
        with open(filename, 'r') as f:
            buff = ''.join([line for line in f])
        self.parse(buff)

    # IS THIS NEEDED? We might just as well print() the bibliography and
    # use `with` context in the caller?
    #
    # def write(self, filename=None, unordered=False):
    #     '''Write the bibliography to a BibTeX file.'''
    #     if not filename and not self.filename:
    #         raise ValueError(filename)
    #     if not filename:
    #         filename = self.filename
    #     with open(filename, 'w') as f:
    #         if unordered:
    #             f.write(repr(self))
    #         # Ensure decent ordering: first preamble, then not-collections,
    #         # then collections—because sometimes BibTeX can’t find crossrefs
    #         # unless they *follow* the reference
    #         else:
    #             f.write(repr(self.by_type('preamble')))
    #             f.write(repr(sorted(self.by_type(['preamble', 'collection'], \
    #                 complement=True))))
    #             f.write(repr(sorted(self.by_type('collection'))))

# Basic test if run as a script
if __name__ == '__main__':
    import sys

    def die(msg):
        print(f'{msg}\n', file=sys.stderr)
        sys.exit(1)

    for name in sys.argv[1:]:
        try:
            db = Biblio(name)
        except FileNotFoundError:
            die(f'File not found: "{name}"')
        except PermissionError:
            die(f'Access denied: "{name}”')
        except IOError:
            die(f'I/O error: "{name}"')
        except ParseError as exc:
            die(f'Parse error on line {exc.lineno}: {exc.err}')
        except DuplicateError as exc:
            die(f'Duplicate ID on line {exc.lineno}: "{exc.err}"')
        except NoIDError as exc:
            die(f'Missing ID on line {exc.lineno}: "{exc.err}"')
        print(db)
